File: 2020/hackceler8/match-2-package/outofgamectf2/runner.cpp
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include <cstddef>
#include <cstdint>
#include <cstdio>

#include <sys/mman.h>
#include <sys/prctl.h>
#include <linux/seccomp.h>
#include <unistd.h>

bool read_into(char *buf, std::size_t to_read) {
  while(to_read) {
    ssize_t nread = read(0, buf, to_read);
    if(nread <= 0)
      return false;
    buf += nread;
    to_read -= nread;
  }
  return true;
}

char *read_data(int prot) {
  uint32_t len = 0;
  if(!read_into(reinterpret_cast<char*>(&len), sizeof(len)))
    return nullptr;

  std::size_t aligned_len = (len + 4096 - 1) & ~(4096 - 1);
  char *data = reinterpret_cast<char*>(mmap(nullptr, aligned_len, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
  if(data == MAP_FAILED)
    return nullptr;
  if(!read_into(data, len))
    return nullptr;
  if(mprotect(data, aligned_len, prot) < 0)
    return nullptr;
  return data;
}

int main() {
  char *data = read_data(PROT_READ | PROT_WRITE);
  char *code = read_data(PROT_READ | PROT_EXEC);

  if(prctl(PR_SET_SECCOMP, SECCOMP_MODE_STRICT))
    return -1;

  typedef int (*fptr)(char*);
  return reinterpret_cast<fptr>(code)(data);
}

File: 2021/quals/misc-shellcode/challenge/runner.py
#!/usr/bin/env python
#
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import random
import string
import subprocess
import sys
import tempfile
import time

def print_to_user(s):
    print(s)
    sys.stdout.flush()

def is_valid(data):
    if len(payload) > 280:
        print_to_user('Payload is too long! It cannot be longer than 280 bytes.')
        return False
    for b in data:
        if b < 0x20 or 0x80 <= b:
            print_to_user('Invalid character 0x%02x found.' % b)
            return False
    return True

def random_string():
    return ''.join(random.choice(string.ascii_lowercase) for i in range(32))

def run_step(binary_command):
    step_flag = random_string()
    with tempfile.TemporaryDirectory() as working_dir:
        with open("%s/flag" % working_dir, "w") as step_flag_file:
            step_flag_file.write(step_flag)
        result = subprocess.run(
                binary_command,
                input=payload,
                cwd=working_dir,
                capture_output=True)
        return result.stdout.strip() == step_flag.encode('utf8')

print_to_user("""
Please provide a payload that prints the contents of 'flag' and runs on both
x86-64 and arm64v8.

The payload may only contain bytes in the range [0x20, 0x7F] and cannot be
longer than 280 bytes.

Payload:
""".strip())

payload = sys.stdin.buffer.readline().strip()
if not is_valid(payload):
    exit()

ok_x86 = run_step(["/home/user/chal-x86-64"])
ok_arm = run_step(["qemu-aarch64", "-L", "/usr/aarch64-linux-gnu", "/home/user/chal-aarch64"])

print_to_user('x86-64 passed:  %s' % ok_x86)
print_to_user('aarch64 passed: %s' % ok_arm)

if ok_x86 and ok_arm:
    with open('/home/user/flag', 'r') as flag_file:
        print_to_user(flag_file.read())
else:
    print_to_user("Sorry, not quite :(")
